File: ios/VineTrack/LegacyImported/Models/PruningActivityListing.swift
import Foundation

/// How a multi-block pruning activity is PRESENTED in a list — one row per
/// parent activity, never one row per block.
///
/// Pure and byte-identical to the Kotlin `PruningActivityListing`, so the
/// Tracker history, the mobile Activity Report and the editor summary can never
/// label the same activity differently.
nonisolated enum PruningActivityListing {

    /// The compact block label of one activity:
    /// * one block  → "Cab Franc"
    /// * two blocks → "Cab Franc + Sauv Blanc"
    /// * three+     → "Cab Franc + Sauv Blanc +2 more"
    static func blockLabel(_ names: [String]) -> String {
        var seen = Set<String>()
        var clean: [String] = []
        for raw in names {
            let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            let name = trimmed.isEmpty ? "Block" : trimmed
            if seen.insert(name).inserted { clean.append(name) }
        }
        if clean.isEmpty { return "No blocks" }
        if clean.count <= 2 { return clean.joined(separator: " + ") }
        return clean.prefix(2).joined(separator: " + ") + " +\(clean.count - 2) more"
    }

    /// True when a free-text search matches the activity: ANY allocation
    /// matching is a match for the whole parent record.
    static func matches(
        query: String,
        blockNames: [String],
        worker: String,
        notes: String,
        rowLabels: [String]
    ) -> Bool {
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !needle.isEmpty else { return true }
        let haystack = (blockNames + [worker, notes] + rowLabels).joined(separator: " ")
        return haystack.localizedCaseInsensitiveContains(needle)
    }

    /// "42–44, 66–67" — contiguous runs of the rows an allocation covered.
    static func rowRangeLabel(_ rows: [Int]) -> String {
        let sorted = Array(Set(rows)).sorted()
        guard let first = sorted.first else { return "—" }
        var parts: [String] = []
        var start = first
        var previous = first
        for row in sorted.dropFirst() {
            if row == previous + 1 {
                previous = row
                continue
            }
            parts.append(start == previous ? "\(start)" : "\(start)–\(previous)")
            start = row
            previous = row
        }
        parts.append(start == previous ? "\(start)" : "\(start)–\(previous)")
        return parts.joined(separator: ", ")
    }
}

/// One quarter a write could not attribute, with the block it belongs to.
nonisolated struct PruningQuarterConflict: Identifiable, Sendable, Hashable {
    let paddockId: UUID?
    var blockName: String = ""
    var row: Int?
    var quarter: Int?
    var reason: String?

    var id: String { "\(paddockId?.uuidString ?? "-")|\(row ?? -1)|\(quarter ?? -1)" }

    var label: String {
        "Row \(row.map(String.init) ?? "?") · q\(quarter.map(String.init) ?? "?")"
    }
}

/// The reconciliation result of ONE activity write (sql/166). A save is never
/// reported as fully successful while conflicts exist: the user is told exactly
/// how many quarters landed, how many were already recorded elsewhere, and which
/// blocks to open to review them.
nonisolated struct PruningActivityReconciliation: Identifiable, Sendable, Equatable {
    let activityId: UUID
    var blockSummary: String = ""
    var quartersRequested: Int = 0
    var quartersRecorded: Int = 0
    var conflicts: [PruningQuarterConflict] = []
    var seasonYear: Int?
    var vintageYear: Int?
    var workTaskConflict: Bool = false
    var isReversal: Bool = false
    var quartersReleased: Int = 0
    var isStale: Bool = false
    var error: String?

    var id: UUID { activityId }

    var quartersConflicted: Int { conflicts.count }

    var hasConflicts: Bool { !conflicts.isEmpty }

    /// Blocks the user can open to review refused quarters.
    var conflictBlockIds: [UUID] {
        var seen = Set<UUID>()
        var ids: [UUID] = []
        for conflict in conflicts {
            guard let id = conflict.paddockId else { continue }
            if seen.insert(id).inserted { ids.append(id) }
        }
        return ids
    }

    func conflicts(in paddockId: UUID) -> [PruningQuarterConflict] {
        conflicts.filter { $0.paddockId == paddockId }
    }

    /// Only a clean, acknowledged write counts as fully synced.
    var isFullySynced: Bool { error == nil && !isStale && !hasConflicts }

    var headline: String {
        if error != nil { return "Activity not saved yet" }
        if isReversal { return "Activity reversed" }
        if hasConflicts { return "Activity saved with conflicts" }
        return "Activity saved"
    }

    var detail: String {
        if error != nil {
            return "Queued — the server hasn't confirmed this activity yet. It will retry automatically."
        }
        if isReversal {
            return "\(quartersReleased) \(Self.quarterWord(quartersReleased)) reopened across \(blockSummary)."
        }
        var parts = ["\(quartersRecorded) \(Self.quarterWord(quartersRecorded)) recorded"]
        if hasConflicts {
            var seen = Set<String>()
            var blocks: [String] = []
            for conflict in conflicts {
                let name = conflict.blockName.isEmpty ? "another block" : conflict.blockName
                if seen.insert(name).inserted { blocks.append(name) }
            }
            parts.append(
                "\(quartersConflicted) \(Self.quarterWord(quartersConflicted)) were already recorded elsewhere (\(blocks.joined(separator: ", ")))"
            )
        }
        if workTaskConflict { parts.append("the Work Task link is already used by another activity") }
        return parts.joined(separator: " · ") + "."
    }

    private static func quarterWord(_ count: Int) -> String {
        count == 1 ? "quarter" : "quarters"
    }

    /// Maps the raw sql/166 response onto what the UI shows. `attributed` is what
    /// actually landed; `conflicts` are quarters another record already owns
    /// (never stolen).
    static func from(
        _ result: PruningActivityResult,
        blockNames: [UUID: String] = [:],
        blockSummary: String = "",
        activityId: UUID,
        isReversal: Bool = false
    ) -> PruningActivityReconciliation {
        let allocationResults = result.allocationResults ?? []
        let allocationConflicts = allocationResults.flatMap { $0.conflicts ?? [] }
        let raw = (result.conflicts?.isEmpty == false) ? (result.conflicts ?? []) : allocationConflicts
        let conflicts = raw.map { conflict in
            PruningQuarterConflict(
                paddockId: conflict.paddockId,
                blockName: conflict.paddockId.flatMap { blockNames[$0] } ?? "",
                row: conflict.row,
                quarter: conflict.segment,
                reason: conflict.reason
            )
        }
        let totals = result.canonical?.totals
        let summary: String = {
            guard let canonical = totals?.blockSummary, !canonical.isEmpty else { return blockSummary }
            return canonical
        }()
        return PruningActivityReconciliation(
            activityId: result.activityId ?? activityId,
            blockSummary: summary,
            quartersRequested: allocationResults.reduce(0) { $0 + ($1.requested ?? 0) },
            quartersRecorded: totals?.quarters ?? allocationResults.reduce(0) { $0 + ($1.attributed ?? 0) },
            conflicts: conflicts,
            seasonYear: result.canonical?.activity?.seasonYear,
            vintageYear: result.canonical?.activity?.vintageYear,
            workTaskConflict: result.workTaskConflict ?? false,
            isReversal: isReversal || (result.reversed ?? false),
            quartersReleased: result.quartersReleased ?? 0,
            isStale: result.stale ?? false,
            error: result.error
        )
    }
}

File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/model/PruningActivityListing.kt
package com.rork.vinetrack.data.model

/**
 * How a multi-block pruning activity is PRESENTED in a list — one row per
 * parent activity, never one row per block.
 *
 * Shared, pure and identical to the iOS `PruningActivityListing`, so the
 * Tracker history, the mobile Activity Report and the editor summary can never
 * label the same activity differently.
 */
object PruningActivityListing {

    /**
     * The compact block label of one activity:
     *  * one block  → "Cab Franc"
     *  * two blocks → "Cab Franc + Sauv Blanc"
     *  * three+     → "Cab Franc + Sauv Blanc +2 more"
     */
    fun blockLabel(names: List<String>): String {
        val clean = names.map { it.trim().ifBlank { "Block" } }.distinct()
        return when {
            clean.isEmpty() -> "No blocks"
            clean.size <= 2 -> clean.joinToString(" + ")
            else -> clean.take(2).joinToString(" + ") + " +${clean.size - 2} more"
        }
    }

    /**
     * True when a free-text search matches the activity: ANY allocation
     * matching is a match for the whole parent record.
     */
    fun matches(
        query: String,
        blockNames: List<String>,
        worker: String,
        notes: String,
        rowLabels: List<String>,
    ): Boolean {
        val needle = query.trim().lowercase()
        if (needle.isEmpty()) return true
        val haystack = (blockNames + worker + notes + rowLabels).joinToString(" ").lowercase()
        return haystack.contains(needle)
    }

    /** "42–44, 66–67" — contiguous runs of the rows an allocation covered. */
    fun rowRangeLabel(rows: List<Int>): String {
        if (rows.isEmpty()) return "—"
        val sorted = rows.distinct().sorted()
        val parts = mutableListOf<String>()
        var start = sorted.first()
        var previous = start
        for (row in sorted.drop(1)) {
            if (row == previous + 1) {
                previous = row
                continue
            }
            parts.add(if (start == previous) "$start" else "$start–$previous")
            start = row
            previous = row
        }
        parts.add(if (start == previous) "$start" else "$start–$previous")
        return parts.joinToString(", ")
    }
}

/** One quarter a write could not attribute, with the block it belongs to. */
data class PruningQuarterConflict(
    val paddockId: String?,
    val blockName: String = "",
    val row: Int? = null,
    val quarter: Int? = null,
    val reason: String? = null,
) {
    val label: String get() = "Row ${row ?: "?"} · q${quarter ?: "?"}"
}

/**
 * The reconciliation result of ONE activity write (sql/166). A save is never
 * reported as fully successful while conflicts exist: the user is told exactly
 * how many quarters landed, how many were already recorded elsewhere, and which
 * blocks to open to review them.
 */
data class PruningActivityReconciliation(
    val activityId: String,
    val blockSummary: String,
    val quartersRequested: Int = 0,
    val quartersRecorded: Int = 0,
    val conflicts: List<PruningQuarterConflict> = emptyList(),
    val seasonYear: Int? = null,
    val vintageYear: Int? = null,
    val workTaskConflict: Boolean = false,
    val isReversal: Boolean = false,
    val quartersReleased: Int = 0,
    val isStale: Boolean = false,
    val error: String? = null,
) {
    val quartersConflicted: Int get() = conflicts.size

    val hasConflicts: Boolean get() = conflicts.isNotEmpty()

    /** Blocks the user can open to review refused quarters. */
    val conflictBlockIds: List<String> get() = conflicts.mapNotNull { it.paddockId }.distinct()

    fun conflicts(paddockId: String): List<PruningQuarterConflict> =
        conflicts.filter { it.paddockId == paddockId }

    /** Only a clean, acknowledged write counts as fully synced. */
    val isFullySynced: Boolean get() = error == null && !isStale && !hasConflicts

    val headline: String
        get() = when {
            error != null -> "Activity not saved yet"
            isReversal -> "Activity reversed"
            hasConflicts -> "Activity saved with conflicts"
            else -> "Activity saved"
        }

    val detail: String
        get() {
            if (error != null) {
                return "Queued — the server hasn't confirmed this activity yet. It will retry automatically."
            }
            if (isReversal) {
                return "$quartersReleased ${quarterWord(quartersReleased)} reopened across $blockSummary."
            }
            val parts = mutableListOf("$quartersRecorded ${quarterWord(quartersRecorded)} recorded")
            if (hasConflicts) {
                val blocks = conflicts.map { it.blockName.ifBlank { "another block" } }.distinct()
                parts.add(
                    "$quartersConflicted ${quarterWord(quartersConflicted)} " +
                        "were already recorded elsewhere (${blocks.joinToString(", ")})"
                )
            }
            if (workTaskConflict) parts.add("the Work Task link is already used by another activity")
            return parts.joinToString(" · ") + "."
        }

    private fun quarterWord(count: Int): String = if (count == 1) "quarter" else "quarters"
}
